crates/driver/src/domain/competition/solution/scoring.rs

```diff
@@ -109,13 +109,15 @@ impl Trade {
                     .0
                     .checked_mul(self.executed.into())?
                     .checked_div(price_limits.buy.0)?;
-                // difference between limit sell and executed amount converted to sell token
-                limit_sell.checked_sub(
-                    self.executed
-                        .0
-                        .checked_mul(self.custom_price.buy)?
-                        .checked_div(self.custom_price.sell)?,
-                )
+                let sold = self
+                    .executed
+                    .0
+                    .checked_mul(self.custom_price.buy)?
+                    .checked_div(self.custom_price.sell)?;
+                // since sell price limit can be lower than order sell limit (e.g. price
+                // improvement fee with quote as price limit), we don't want to
+                // return error in cases when the solution is worse than price limit (quote)
```

Comment:
Would we want to return an error if we have a negative surplus compared to the actual limit price of the order (not just the quote)?

Also I find the comment quite hard to understand.
IIUC the heart of the problem is that that `surplus` may be called with different price limits depending on the fee (order limit price vs. quoted price). And there is the option that an execution respects the limit price but does not actually improve over the quoted price. In that case this function would return an error without this patch, right?

Reply:
> Would we want to return an error if we have a negative surplus compared to the actual limit price of the order (not just the quote)?

No. We also do not return error when calculating the protocol fee to be withheld (fee.rs). Here, it is even unexpected to ever happen since scoring is called after solution passed simulation (although we should not assume wider context of function usage to decide on these kind of things).

>  In that case this function would return an error without this patch, right?

Yes

Reply:
Used `surplus_over_reference_price` as we already have a function with similar functionality.
Introduced `surplus_over_limit_price` and `surplus_over_quote`. 
Updated the comments.

Can you check if it is clearer now?